include/proxy-wasm/context_interface.h

```diff
@@ -0,0 +1,628 @@
+  /**
+   * Called on Root Context to with the initial metadata of a grpcStream.
+   * @token is the token returned by grpcStream.
+   * The metadata can be retrieved via the HeaderInterface functions with
+   * WasmHeaderMapType::GrpcReceiveInitialMetadata as the type.
+   */
+  virtual void onGrpcReceiveInitialMetadata(GrpcToken token) = 0;
+
+  /**
+   * Called on Root Context when gRPC data has arrived.
+   * @token is the token returned by grpcCall or grpcStream.
+   * For Call this implies OK close.  For Stream may be called repeatedly.
+   * The data can be retrieved via the getBuffer function with
+   * WasmBufferType::GrpcReceiveBuffer as the type.
+   */
+  virtual void onGrpcReceive(GrpcToken token) = 0;
+
+  /**
+   * Called on Root Context to with the trailing metadata of a grpcStream.
+   * @token is the token returned by grpcStream.
+   * The trailers can be retrieved via the HeaderInterface functions with
+   * WasmHeaderMapType::GrpcReceiveTrailingMetadata as the type.
+   */
+  virtual void onGrpcReceiveTrailingMetadata(GrpcToken token) = 0;
+
+  /**
+   * Called on Root Context when a gRPC call has closed prematurely or a gRPC stream has closed.
+   * @token is the token returned by grpcStream.
+   * @status is status of the stream.
+   * @message is any status message.
+   * For Call only used when not Ok. For stream indicates the peer has closed.
+   */
+  virtual void onGrpcClose(GrpcToken token, const GrpcStatus &status,
+                           const string_view message) = 0;
+};
+
+/**
+ * RootInterface is the interface specific to RootContexts.
+ * A RootContext is associated with one more more plugins and is the parent of all stream Context(s)
+ * created for that plugin. It can be used to store data shared between stream Context(s) in the
+ * same VM.
+ */
+struct RootInterface : public RootGrpcInterface {
+  /**
+   * Call on a host Context to create a corresponding Context in the VM.  Note:
+   * onNetworkNewConnection and onRequestHeaders() call onCreate().
+   * @param parent_context_id is the parent Context id for the context being created.  For a
+   * stream Context this will be a Root Context id (or sub-Context thereof).
+   */
+  virtual void onCreate(uint32_t parent_context_id);
+
+  /**
+   * Call on a Root Context when a VM first starts up.
+   * @param plugin is the plugin which caused the VM to be started.
+   * Called by the host code.
+   */
+  virtual bool onStart(std::shared_ptr<PluginBase> plugin) = 0;
+
+  /**
+   * Call on a Root Context when a plugin is configured or reconfigured dynamically.
+   * @param plugin is the plugin which is being configured or reconfigured.
+   */
+  virtual bool onConfigure(std::shared_ptr<PluginBase> plugin) = 0;
+
+  /**
+   * Called on a Root Context when a response arrives for an outstanding httpCall().
+   * @param token is the token returned by the corresponding httpCall().
+   */
+  virtual void onHttpCallResponse(HttpCallToken token) = 0;
```

Comment:
This seems like the odd one out, why isn't it in something like `RootHttpInterface`? Analogous to `RootGprcInterface`.